cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/PluginRepository.java

```diff
@@ -0,0 +1,479 @@
+import com.google.common.base.Supplier;
+import com.google.common.base.Throwables;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+import com.google.common.io.Closeables;
+import com.google.common.io.Files;
+import com.google.common.primitives.Primitives;
+import com.google.common.reflect.TypeToken;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.io.Reader;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.jar.JarFile;
+import javax.annotation.Nullable;
+
+/**
+ * This class manage plugin information that are available for application templates.
+ */
+public class PluginRepository {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PluginRepository.class);
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(PluginClass.class, new PluginClassDeserializer())
+    .create();
+
+  // Object type for data in the config json file.
+  private static final Type CONFIG_OBJECT_TYPE = new TypeToken<List<PluginClass>>() { }.getType();
+
+  // Transform File into PluginInfo, assuming the plugin file name is in form [name][separator][version].jar
+  private static final Function<File, PluginFile> FILE_TO_PLUGIN_FILE = new Function<File, PluginFile>() {
+    @Override
+    public PluginFile apply(File file) {
+      String plugin = file.getName().substring(0, file.getName().length() - ".jar".length());
+      PluginVersion version = new PluginVersion(plugin, true);
+      String rawVersion = version.getVersion();
+
+      String pluginName = rawVersion == null ? plugin : plugin.substring(0, plugin.length() - rawVersion.length() - 1);
+      return new PluginFile(file, new PluginInfo(file.getName(), pluginName, version));
+    }
+  };
+
+  private final CConfiguration cConf;
+  private final File templateDir;
+  private final File tmpDir;
+  private final AtomicReference<Map<String, Multimap<PluginInfo, PluginClass>>> plugins;
```

Comment:
Map from template name to all plugins available. The `PlugInfo` is basically information about the plugin file (e.g. transform-1.0.jar). Inside a jar, there can be multiple plugin classes (e.g. Identity, Projection, etc), hence a multimap.